[Nguyen Quang Vinh] iP

IceWizard4902 opened the pull request:
## DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is

1.  download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html)
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 😉 

And it is **FREE**!
 
Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

If you Java programmer, you can use it to practice Java too. Here's the `main` method
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

IceWizard4902 committed:
Add basic greetings to the program

IceWizard4902 committed:
Add functionalities for the skeletal version of Duke in Level 1

IceWizard4902 committed:
Add the ability to store tasks and display to the user when requested

IceWizard4902 committed:
Add the ability to mark tasks as done.
Add classes:
* Task: to represent an individual task
* TaskList: to represent the task list of the user
Make changes to the original code to adapt to the changes.
To-do: Fixing errors in corner cases, bug squashing

IceWizard4902 committed:
Add support for the three types of tasks, refractor code

IceWizard4902 committed:
Use the input/output redirection technique to semi-automate the testing of Duke

IceWizard4902 committed:
Add additional test for exit feature of the app

IceWizard4902 committed:
Add javadocs to new classes of tasks

IceWizard4902 committed:
Add exceptions to Duke as well as logic to detect invalid inputs.
Perform code refractoring.
Improve input parsing. Instead of splitting the string by every space in the input, we first split the input into two parts separated by the first space,
then for deadline and event tasks we split by their respective signal (/at, /by)
Include responses in methods provided by TaskList.

IceWizard4902 committed:
Javadocs addition

IceWizard4902 committed:
Add support for deleting tasks from the list.
Code adaptation to storing the task using Java collection.

IceWizard4902 committed:
Add feature to save the tasks in the hard disk automatically whenever the task list changes

IceWizard4902 committed:
Add feature to help Duke understand dates

IceWizard4902 committed:
Resolve conflicts

IceWizard4902 committed:
Code changes to adapt to the new feature implemented in Level-8

IceWizard4902 committed:
Refractor code to make the code more OOP

IceWizard4902 committed:
Organize the classes into suitable java packages.

IceWizard4902 committed:
Add JUnit tests

IceWizard4902 committed:
Add and change Javadocs of classes

IceWizard4902 committed:
Adapt the code to follow the given standards

IceWizard4902 committed:
Add a feature to give users a way to find a task by searching for a keyword

IceWizard4902 committed:
Resolve conflicts

IceWizard4902 committed:
Resolve conflicts

IceWizard4902 committed:
Code and docs clean-up

vimuthm commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+import tasks.ToDo;
+import tasks.Deadline;
+import tasks.Event;
+import tasks.TaskList;
```
Alphabetically ordered imports might increase readability!

vimuthm commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+ * Input parsing and displaying output to the user.
+ */
+public class Parser {
+    private final Scanner sc;
```
It might make sense to handle `Scanner` object creation and handling within the `Ui` class. Thoughts?

Yttruire replied:
I think I agree with Vimuth. We might naturally expect a `Parser` class to only parse input and sort them out/respond to input that cannot be parsed, but not necessarily scanning for the input itself. The `Ui` class on the other hand, could definitely handle both input (receiving input) and output (printing messages to user)

vimuthm commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @param tasks The task list to be saved.
+     */
+
```
There shouldn't be a line between the ending javadoc and function declaration.

vimuthm commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+    public static ArrayList<Task> loadTaskListFromHardDisk() throws IOException {
+        File dir = new File(DIRECTORY);
+        File logs = new File(FILEPATH);
+        ArrayList<Task> result = new ArrayList<>();
```
Collections variables should be named with plural words

vimuthm commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Ui {
+
+    public static void greetings() {
```
This public method would need javadocs

vimuthm commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,61 @@
+        this.isDone = false;
+    }
+
+    public String getDescription() {
```
This would require javadoc

vimuthm commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     * Add a task to the list
+     *
+     * @param task The added task.
+     */
```
This comment would require the return type.

vimuthm commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return Size of the task list.
+     */
+    public int size() {
```
This method might be better named `getSize`

vimuthm commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+            StringBuilder res = new StringBuilder("Here are the tasks in your list:\n");
+            for (int i = 0; i < tasks.size(); i++) {
+                String temp = "\t" + " " + (i + 1) + "." + tasks.get(i).toString();
+                if (i < tasks.size() - 1) { // remove the last \n char, ugly but get the job done
```
Nice :) I handled this similarly!

vimuthm commented on src/test/java/duke/StorageTest.java:
```diff
@@ -0,0 +1,30 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+import tasks.*;
```
Wildcard imports should not be used

vimuthm reviewed (commented):
Overall, it looks quite good and nice use of OOP!

One thing I would point out is that the `Parser` class is also handling the user input from the scanner which might be better suited for the `Ui` class. 

A `Command` class would also be a nice addition if time permits.

Other than that, it looks good to me!

Yttruire commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,13 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        Ui.greetings();
```
Method names should be verbs instead of nouns, according to the coding standard that we are following. Instead, consider something like `Ui.sendGreetings()` or `Ui.greetUser()`

Yttruire commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+        Ui.displayContentBetweenLines(taskList.findTask(parsedInput[1]));
+    }
+
+    private void dukeCommandController(String[] parsedInput) throws DukeInvalidCommandException {
```
Notice that this method is a noun, and actually, sounds like an object. Perhaps this would be better as an individual class for the sake of OOP if you have time. Else, the actions to take in response to a command you retrieve from parsing, may fit better elsewhere (Maybe in the main class if you're not encapsulating this into a different class) rather than leave this in the `Parser`

Yttruire commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @param content Content to display.
+     */
+    public static void displayContentBetweenLines(String content) {
```
This method name sounds a little verbose, although there is nothing wrong with it. Perhaps try names that describe what  it does less, and more of what it is for? E.g. If this is the function to be used to display every form of output, this could simply be named `display` or `displayMessage`.

Yttruire commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+    protected LocalDate by;
+
+    /**
+     * Constructor for the class.
```
The first sentence as a summary for the method should be a verb to be as descriptive as possible. According to the coding standards, it could start with `Returns...`, `Sends...`, or `Adds...` for example. Thus, I think that `Constructs` might be more appropriate here.

Yttruire commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,61 @@
+     * @return The string representation.
+     */
+    public String toSaveInHardDisk() {
+        return "";
```
If this method is not intended to be useful as it is, perhaps consider making this and the class, an abstract method and abstract class responspectvely

Yttruire commented on src/test/java/duke/StorageTest.java:
```diff
@@ -0,0 +1,30 @@
+
+public class StorageTest {
+    @Test
+    public void checkIfSavedFileExist() {
```
Since this is a test rather than a simple boolean check (if a certain file exists), and is instead a test for whether a save file can be saved and reloaded, I think it can be named more appropriately. E.g. `saveFileTest` perhaps?

Yttruire commented:
Everything looks great, and I think that it is an overall job well done! There are no major issues, and the code base was easy to understand. There are just some small pointers that we have both brought up.

IceWizard4902 commented:
Thanks for the reviews! Really appreciate you guys going through my code!

IceWizard4902 committed:
Change the code from remarks in reviews

IceWizard4902 committed:
Merge branch 'add-gradle-support'

IceWizard4902 committed:
Add Checkstyle support

IceWizard4902 committed:
Refractor code based on reviews given

IceWizard4902 committed:
Improve code from reviews given

IceWizard4902 committed:
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

IceWizard4902 committed:
Add Java assertions

IceWizard4902 committed:
Merge pull request #1 from IceWizard4902/branch-A-Assertions

Merge branch branch-A-Assertions

IceWizard4902 committed:
Coding quality: improve the quality of code by following coding standards

The current code uses magic literals, has little use of DRY. The methods in Parser have some common behaviors.

The current code quality causes code duplication and low readability.

Extracting the common behavior into a method allows us to avoid duplication.
Replacing magic literals with constants.
Improving parsing so that the toString method in Task does not have to convert the date into the correct format.

Following code standards in CS2103 here is ideal.

Code improvements are inspired by repositories on Github.

IceWizard4902 committed:
Merge pull request #2 from IceWizard4902/branch-A-CodeQuality

Improve code quality

IceWizard4902 committed:
Code quality: Improve the quality

There are still magic literals in the Parser code. Some of the refractorings previously lead to wrong response by the taskList.

Magic literals for command name are replaced by enum.
Amend the changes to the wrong part due to previous refractoring.

Let's find more ways to improve the code by reading other repositories.

IceWizard4902 committed:
Provide a way to tag items

IceWizard4902 committed:
Add credits to the code. Squash bugs

IceWizard4902 committed:
Add a user guide

IceWizard4902 committed:
Set theme jekyll-theme-cayman

IceWizard4902 committed:
Fix the error in the image file

IceWizard4902 committed:
Merge branch 'master' of https://github.com/IceWizard4902/ip

IceWizard4902 committed:
Add exit feature when the user type the bye command

IceWizard4902 committed:
Fix name for image
